### DISCUSSION TITLE
good UX requires a GUI

### DISCUSSION THREAD

**Discussion by nuuSolutions** [typst/typst#2858] [Polls]
I highly respect the effort to clean up the LaTeX mess.
Both TeX and Typst would, however, greatly benefit from a well designed GUI.
(GUI is apparently hard. There is LyX, but unfortunately it never got out of the hobby project phase)

**Comment by Enivex** [comment]
> good UX requires a GUI

I don't agree with your premise

**Reply by Dherse** [reply]
Tools like Vivado and Quartus are proof that a GUI isn't necessarily better.

**Comment by nuuSolutions** [comment]
Quite the expected result so far.
If we find one person interested in a GUI I'm going for it - found my niche market then

**Reply by Dherse** [reply]
What do you mean niche market?

**Reply by maribox** [reply]
I would find something like [marktext](https://github.com/marktext/marktext) except for Typst amazing. This would enable one to have a document on the left side of your screen and a Typst document on the right, so you don't have to have two Monitors for taking notes while reading for example, as you currently need to have with Typst, because it's [YAFIYGI](http://catb.org/jargon/html/Y/YAFIYGI.html) and there's currently no WYSIWYG Typst editor afaik

**Reply by alerque** [reply]
There is nothing stopping 3rd party projects from doing this. In fact just such apps already exist. See for example [typstudio](https://github.com/Cubxity/typstudio). Also you don't have to have two monitors to have an editor window and PDF viewer open side by side. If you leave the compiler running in the background watching for file changes and use a PDF viewer that automatically updates the experience is pretty good and you get to use your favorite editor.

Some editors also have preview plugins available that render in a side by side pane already too.

**Reply by maribox** [reply]
Sorry, I meant a different pdf that I read on one side of the monitor and then can take notes on the other side of the monitor. That's only possible with a WYSIWYG editor, which typstudio isn't, right?

**Comment by UARTman** [comment]
As LyX has been highly helpful to me, allowing me to do a number of math assignments *entirely* in LaTeX (as opposed to solving it on paper and then painstakingly typesetting the results), I fell like there's at least some potential for work in this direction. Ironically, I think Typst's structure might lend itself to LyX-like workflow, what with blocks and stuff... I am seriously thinking of doing some hobby-level exploration in that direction, though I'm not likely to make it actually work.

**Reply by nuuSolutions** [reply]
Yeah, I use LyX in a similar way, but the experience could really be much better.
There also was Scientific Workplace, I wonder how much sales they had - not too much probably as they are out of business.
I also tried Mathematica some time ago, but there have been other quirks, maybe I'll try again

**Reply by Doublonmousse** [reply]
The issue is Lyx is you gain a much better way to fix errors (something that is less of an issue with typst as you can see changes as soon as you do them) but you end up with limitations when typing math (sometimes it's actually more flexible to type things than using it the gui, when you know the command but not the way to do it with the gui or when the gui way is actually impeding writing), and there are other weird things as far as the UX goes in lyx (having a separate type of file and sometimes weird looking tex files). Basically you get a GUI-like experience but you loose any way to tweak the tex code.

I do think some gui can be useful but you have to be careful with design. You could do a word-like interface for typst (for text editing and images, but to edit equations I don't see how you could do it without showing the related code) but I can't see it any other way than a separate preview mode. I feel it would be a much better UX to have it as an option, with the other workflow accessible as well at the same time. I don't think all of the things you can do with typst (the actual program-oriented things) can translate to a GUI interface or if this would even be desirable.

**Comment by basings** [comment]
do you mean something like [WYSIWYG](https://en.wikipedia.org/wiki/WYSIWYG)?

if not. what is wrong with typstudio or vscodium or pulsar? both have top notch UI

**Reply by davidleejy** [reply]
Tried typstudio a few weeks ago and couldn't get it to work directly out of the box. I don't blame the developers as the project is, as stated by them, a work in progress (WIP). 

Currently am using VSCode with two extensions for Typst – Typst Preview and Typst LSP.

**I am happy to share that this set-up**:

1. Has "**live compilation**". Results of edits are shown in real-time in a separate VSCode tab that shares screen real estate with the tab showing the typ file I'm editing. (_Thanks to Typst Preview_)

2. Is able to let me **jump to content** in pdf file being previewed (see point 1) when the associated content is clicked in the tab showing the typ file. This jump also goes the "other way" – by clicking on content in the pdf preview, the tab showing the typ file scrolls to the related part. (_Thanks to Typst Preview_)

3. **Automatically saves the resulting pdf file** whenever the typ file being edited is saved. (_Thanks to Typst LSP_)
 
**Problems**:

Happy to report no major issues, and am pretty satisfied with the smoothness of this set-up thus far. ⛵️🌊🌊

One minor issue that occurs from time to time, and especially after sleeping the machine or leaving the machine idle, is the need to "re-engage" Typst Preview due to it having stopped watching the typ files that it's supposed to be watching. Luckily, resolving this is straightforward and quick. Ctrl + shift + P in VSCode to pop up the "omnibar" (sorry "omnibar" is likely not standard VSCode terminology), typing "preview" to surface the command "Typst Preview: Preview Current File", and finally pressing enter to execute it.

**Comment by nuuSolutions** [comment]
In an ideal UX case you would be able to write in markup AND edit in the output view

**Reply by johanneskoester** [reply]
> In an ideal UX case you would be able to write in markup AND edit in the output view

This is exactly what would be needed. Typst is very nice like it is intended now, and completely fine for people with computational backgrounds. However, at least in academia, one often has to collaborate with non-computational people, and many of them will be overwhelmed by the syntactical complexity of WYSIWYM. For those cases, it would be invaluable for them to be able to edit and comment on the right hand side, in the preview panel. Importantly, this does not require them to be able to edit everything. For example the math can be left read-only. If they can edit the text parts and also extend them, that would be enough. People who are capable to understand and edit the math will also have no issues to master the typst syntax.

Like the swiftlatex editor, but without aparently discontinued development and in modern.

**Reply by fenke** [reply]
Posit's Positron - their follow-up to R-Studio based on VSCode - your text is basically entered as plain-text but it also has a visual editor mode and - iirc - VSCode standard markdown editor features a live preview in which you can edit.

It looks like a mixed mode typst editor should be possible?

**Comment by HPDell** [comment]
I think what we need is something called "instant rendering" like Typora instead of a GUI editor.

**Comment by felidz** [comment]
I am a LyX user, too. I believe many who prefer markdown+script over GUI might have less experience of typing heavy mathematical formulas. I think the advantages and benefits of GUI would be recognized if they had to routinely type and edit something like this. 
![UoXmW](https://github.com/typst/typst/assets/24916531/a12f6002-4cf3-4228-a556-f3b1aeab51b0)

Even Obsidian (operating in a markdown+script way), which has a community plugin called LaTeX Suite—a major reason many prefer it over Logseq—suffers from less than ideal experiences. It often experiences flickering issues; for instance, when you click on a formula to edit it, it doesn’t always take you to the expected position, especially with longer formulas.

The math feature in GUI is the only feature I truly care about, since LyX has tripled my speed in typing math compared to other tools, which tend to have more friction. And I would advocate for the promotion of GUI based on this utility alone.

**Reply by davidleejy** [reply]
I'm on the same page with felidz about the possibility of a GUI being very helpful for users who have to edit complex mathematical content.

It would be very convenient to have the ability to click on a part of a rendered mathematical formula and have the editor's caret be transported to the Typst code responsible for it. For a start, even slightly inaccurately transporting the caret to the vicinity of the Typst code would be helpful.

**Reply by laurmaedje** [reply]
> It would be very convenient to have the ability to click on a part of a rendered mathematical formula and have the editor's caret be transported to the Typst code responsible for it. For a start, even slightly inaccurately transporting the caret to the vicinity of the Typst code would be helpful.

That already exists in the web app, and I think in Typst LSP as well.

**Reply by felidz** [reply]
Perhaps a counterpart to a more robust version of https://cortexjs.io/mathlive/ could be a solution.

**Reply by fbob** [reply]
I second what @felidz  suggest : mathlive is a good exemple of the use of a  gui to edit complex math equations.

**Reply by atbug** [reply]
I actually like TeXmacs style editting more, although it is built on a technology that I don't like. TeXmacs has convenient shortcuts for common math symbols. LyX is also a good alternative.

**Reply by nuuSolutions** [reply]
> I am a LyX user, too. I believe many who prefer markdown+script over GUI might have less experience of typing heavy mathematical formulas.

yeah, that's basically what I was after.
For basic stuff markdown is obviously fine. For utmost generality a scripting language is great.

For devs with a CS background (how likely is that?) all they probably care about is having code inside of the main text, which is trivial with markdown.

But all approaches struggle with editing math and tables. Not to mention math inside a table ;-)
The existence of "equation editors" is a farce pointing at this issue from the other direction.

All WE want is to write text WITH math! In a convenient way.

**Comment by bluebear94** [comment]
Plain text is okay for me, but one area where I might prefer GUIs is typesetting tables, especially ones with many columns (such as phoneme charts).

**Comment by spacekookie** [comment]
I'm a prose/ fiction writer and curious about using typst to store my manuscripts. And while I am also a programmer, I don't like writing fiction in a code editor. This is definitely a very personal thing, but Markdown just doesn't work for me. Having to write random symbols for formatting takes me out of the writing flow (`Ctrl+Shift+Right arrow` + `Ctrl+i` is more convenient :shrug:)

I know writing a wysiwyg editor is a bunch of work, but it would still be appreciated.

**Reply by alerque** [reply]
There are plenty of Markdown editors that are not "code editors" and support <kbd>Ctrl</kbd><kbd>I<kbd> for Italic for example. Some hide the formatting marks, others show them but give you shortcuts for creating them. There are even plenty of them geared specifically at *writers* not coders.

Obviously not many of those exist yet for Typst (although conversion via `pandoc` is of course an option), I'm mostly just pointing out the category error and that there is a whole world of editors in between "code editors" and WYSIWYG, and as a prose author I would suggest the best way to not get lost in either the code or the exact output formatting while writing is somewhere in-between.

**Reply by spacekookie** [reply]
I've tried a few markdown editors before that are a bit more wysiwyg-ish. But they all kinda sucked. To be fair though, it's been a few years so maybe I should revisit the topic.

Currently I'm writing in LibreOffice, which is... fine. It's very slow on my wayland laptop. But I also just noticed that trying to combine different documents (one for each scene) with `pandoc` strips all my in-line formatting (i.e. done the way I described in my OP, instead of setting a paragraph style).

That + `.odt` files being very annoying to keep versioned (and .fodt files being such a fire hose of diff for every save) made me curious to use typst as a storage format, which I can then generate `.odt` files (to send to my editor) and whatever else from when the time comes.

Anyway, I know writing UIs is a lot of work (too much work). I wanted to weigh in my opinion/ situation in case anyone with the know-how desires to feel nerd sniped

**Comment by charlesroddie** [comment]
There are big benefits of text-based formats. Source control and AI collaboration.

But there are also big costs compared to UIs:
- The ordinary human user is not going to edit syntax. The only way that text-based formats get any penetration beyond programmers is via AI where the instructions are given in a chat box and the AI does the edits.
- You see everything twice, in text and in preview. You need two screens to work.

I'm doing a lot of migrations to typst and am going to try this on other people.

The costs are minor for me because I'm a programmer so don't mind editing text occasionally, and I usually work with 2 monitors.

But for non-programmers, the experience of not being able to edit a document normally may be a blocker. If the approach is AI-only for editing, that's very productive sometimes, but also expensive and not completely reliable.

**Comment by nuuSolutions** [comment]
Full WYSIWYG UI expected in 2087 if we can keep this pace up!

Jokes aside:
At the time of my original post I did not even realize that a typst playground website existed.
Which obviously is a GUI but the experience could still be improved massively.

Meanwhile I'm using Overleaf which is also not perfect but has a "Visual Editor" mode, a kind of in-between or mix of WYSIWYG and code. 
I think that's quite a good compromise. Math is rendered in place but can be edited on click.
